controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,369 @@
+)
+
+const (
+	HTTPRouteParents = ".metadata.parent"
+)
+
+// RateLimitingWASMPluginReconciler reconciles a WASMPlugin object for rate limiting
+type RateLimitingWASMPluginReconciler struct {
+	reconcilers.TargetRefReconciler
+}
+
+//+kubebuilder:rbac:groups=extensions.istio.io,resources=wasmplugins,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=httproutes,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=ratelimitpolicies,verbs=get;list;watch;update;patch
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
+func (r *RateLimitingWASMPluginReconciler) Reconcile(eventCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger().WithValues("Gateway", req.NamespacedName)
+	logger.Info("Reconciling rate limiting WASMPlugin")
+	ctx := logr.NewContext(eventCtx, logger)
+
+	gw := &gatewayapiv1.Gateway{}
+	if err := r.Client().Get(ctx, req.NamespacedName, gw); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("no gateway found")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get gateway")
+		return ctrl.Result{}, err
+	}
+
+	if logger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(gw, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info(string(jsonData))
+	}
+
+	desired, err := r.desiredRateLimitingWASMPlugin(ctx, gw)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileResource(ctx, &istioclientgoextensionv1alpha1.WasmPlugin{}, desired, rlptools.WASMPluginMutator)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Rate limiting WASMPlugin reconciled successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) desiredRateLimitingWASMPlugin(ctx context.Context, gw *gatewayapiv1.Gateway) (*istioclientgoextensionv1alpha1.WasmPlugin, error) {
+	wasmPlugin := &istioclientgoextensionv1alpha1.WasmPlugin{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WasmPlugin",
+			APIVersion: "extensions.istio.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rlptools.WASMPluginName(gw),
+			Namespace: gw.Namespace,
+		},
+		Spec: istioextensionsv1alpha1.WasmPlugin{
+			Selector:     common.IstioWorkloadSelectorFromGateway(ctx, r.Client(), gw),
+			Url:          rlptools.WASMFilterImageURL,
+			PluginConfig: nil,
+			// Insert plugin before Istio stats filters and after Istio authorization filters.
+			Phase: istioextensionsv1alpha1.PluginPhase_STATS,
+		},
+	}
+
+	pluginConfig, err := r.wasmPluginConfig(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	if pluginConfig == nil || len(pluginConfig.RateLimitPolicies) == 0 {
+		common.TagObjectToDelete(wasmPlugin)
+		return wasmPlugin, nil
+	}
```

Comment:
nit: `nil` wasmplugin seems to be only returned when `len(wasmPlugin.RateLimitPolicies) == 0` already at 

https://github.com/Kuadrant/kuadrant-operator/blob/e76cd28f7ee1633714a4e853f260e976f8e018f9/controllers/ratelimitpolicy_istio_wasmplugin.go#L231-L234

so I think there's no need to check this again here 🤔 

```suggestion
	if pluginConfig == nil {
		common.TagObjectToDelete(wasmPlugin)
		return wasmPlugin, nil
	}
```

Reply:
I think this comment no longer applies to the current codebase in this PR. Let me know if it still does.